File: src/Patches/Contribution.php
<?php

declare(strict_types=1);

namespace Upkeep\Patches;

use Upkeep\Drupal\Issue;
use Upkeep\Drupal\IssueReference;
use Upkeep\Gitlab\MergeRequest;

/**
 * One drupal.org issue together with every merge request that claims
 * authorship of it — the unit `upkeep patches` classifies and renders.
 *
 * Both halves are kept because an issue can hold both kinds of contribution
 * at once, which is normal in the Drupal community: a patch posted in comment
 * 4, an MR opened in comment 9, and a re-roll posted in comment 14 that never
 * made it onto the branch.
 */
final readonly class Contribution
{
    /**
     * @param list<MergeRequest> $mergeRequests every MR whose metadata asserts
     *                                          authorship of this issue
     *                                          (IssueReference::extractOwning)
     */
    public function __construct(
        public string $module,
        public Issue $issue,
        public array $mergeRequests = [],
    ) {
    }

    /**
     * Pair each issue with the merge requests that claim authorship of it.
     *
     * Shared by every consumer that has to decide what a contribution *is* —
     * the patch report and the dashboard — so the two can never disagree about
     * whether an issue is covered. Merge requests claiming an issue outside
     * $issues are dropped: they cannot change any row, and dropping them is
     * what bounds the emptiness probing the caller may do.
     *
     * @param list<Issue>        $issues
     * @param list<MergeRequest> $mergeRequests
     * @param array<int, int>    $forkNids source-project id => issue nid, from
     *                                     GitlabClient::issueForkNids(). The
     *                                     authoritative pairing where it has
     *                                     an answer: drupal.org made that fork
     *                                     *for* that issue, which is a fact
     *                                     about how the repository exists
     *                                     rather than a string in a title.
     * @return list<self>
     */
    public static function pair(string $module, array $issues, array $mergeRequests, array $forkNids = []): array
    {
        $wanted = [];
        foreach ($issues as $issue) {
            $wanted[$issue->nid] = true;
        }

        $byNid = [];
        foreach ($mergeRequests as $mr) {
            // The fork wins. It is the only thing that pairs a Project Update
            // Bot MR at all: those are titled "Automated Project Update Bot
            // fixes" on a branch called project-update-bot-only, and say only
            // "Relates to #NNN" — which extractOwning() rejects by design.
            $nid = ($mr->sourceProjectId !== null ? ($forkNids[$mr->sourceProjectId] ?? null) : null)
                ?? IssueReference::extractOwning($mr->title, $mr->sourceBranch, $mr->description);

            if ($nid === null || !isset($wanted[$nid])) {
                continue;
            }
            $byNid[$nid][] = $mr;
        }

        $contributions = [];
        foreach ($issues as $issue) {
            $contributions[] = new self($module, $issue, $byNid[$issue->nid] ?? []);
        }

        return $contributions;
    }

    /**
     * The merge request whose work has landed, if one has.
     *
     * The question this whole surface exists to answer for a class of issue
     * that cannot answer it itself: Project Update Bot compatibility issues
     * are kept open on purpose, so the bot can post again as core moves, and
     * an open one may already have had its work merged months ago. Two such
     * issues look identical until you ask whether anything was merged.
     */
    public function landed(): ?MergeRequest
    {
        return MergeRequest::latestMerged($this->mergeRequests);
    }

    /**
     * Whether anything on the issue is newer than the merge.
     *
     * The discriminator, and the reason this never says "resolved". A bot that
     * posts again after its earlier work merged has raised new work; an issue
     * where nothing has happened since is one whose open status is only the
     * convention. Both are true statements about evidence, and which of them
     * warrants closing the issue stays the maintainer's call.
     */
    public function hasWorkNewerThanLanding(): bool
    {
        $landed = $this->landed();
        if ($landed?->mergedAt === null) {
            return false;
        }

        $mergedAt = strtotime($landed->mergedAt);
        if ($mergedAt === false) {
            return false;
        }

        foreach ($this->issue->files as $file) {
            if ($file->timestamp > $mergedAt) {
                return true;
            }
        }

        foreach ($this->mergeRequests as $mr) {
            if ($mr->state === 'merged' || $mr->updatedAt === null) {
                continue;
            }
            $updated = strtotime($mr->updatedAt);
            if ($updated !== false && $updated > $mergedAt) {
                return true;
            }
        }

        return false;
    }

    /**
     * The merge requests that carry changes.
     *
     * An MR whose emptiness is *unknown* counts as substantive. Unknown is the
     * reading a merge-request list payload gives (it omits diff_refs) and the
     * reading a snapshot cached by an older upkeep gives, and in both cases
     * the safe direction is the one that preserves the pre-existing behaviour
     * — treat the MR as real work — rather than one that invents empty MRs
     * out of missing data.
     *
     * @return list<MergeRequest>
     */
    public function substantiveMergeRequests(): array
    {
        return self::substantive($this->mergeRequests);
    }

    /**
     * @param list<MergeRequest> $mergeRequests
     * @return list<MergeRequest>
     */
    public static function substantive(array $mergeRequests): array
    {
        return array_values(array_filter(
            $mergeRequests,
            static fn (MergeRequest $mr): bool => $mr->carriesChanges() !== false,
        ));
    }

    public function kind(): ContributionKind
    {
        $hasPatches = $this->issue->patchCount() > 0;
        $substantive = $this->substantiveMergeRequests() !== [];

        if (!$hasPatches) {
            return $substantive ? ContributionKind::MergeRequestOnly : ContributionKind::Nothing;
        }
        if ($substantive) {
            return ContributionKind::PatchAndMergeRequest;
        }

        return $this->mergeRequests === []
            ? ContributionKind::PatchOnly
            : ContributionKind::PatchWithEmptyMergeRequest;
    }

    /**
     * The revision a cached check result must name to be current for this
     * issue: the newest patch on it. Null when the issue carries no patch at
     * all, in which case there is nothing a result could be about.
     */
    public function currentRevision(): ?string
    {
        $latest = $this->issue->latestPatch();

        return $latest === null ? null : PatchRevision::of($latest->url);
    }

    /**
     * The dashboard's STATUS cell for a patch row: what arrived, and how much
     * of it. Deliberately not a gate verdict — nothing here is mergeable, and
     * a cell that looked like one would invite the wrong action.
     */
    public function dashboardStatus(): string
    {
        $count = $this->issue->patchCount();
        $files = $count === 1 ? '1 patch' : $count . ' patches';

        return match ($this->kind()) {
            ContributionKind::PatchOnly => 'PATCH ' . $files,
            ContributionKind::PatchAndMergeRequest => 'PATCH ' . $files . ', ' . $this->mergeRequestCell(),
            ContributionKind::PatchWithEmptyMergeRequest => 'PATCH ' . $files . ', ' . $this->mergeRequestCell(),
            ContributionKind::Nothing => 'PATCH nothing attached',
            ContributionKind::MergeRequestOnly => 'PATCH covered by ' . $this->mergeRequestCell(),
        };
    }

    /**
     * The MR column: the representative merge request, flagged when it carries
     * nothing, with a count of any others. A substantive MR represents the
     * issue in preference to an empty one — an issue can carry both a real
     * branch and a bot's empty draft, and the real branch is the answer to
     * "is this already in git?".
     */
    public function mergeRequestCell(): string
    {
        return self::renderMergeRequestCell(
            $this->mergeRequests,
            $this->landed(),
            $this->hasWorkNewerThanLanding(),
        );
    }

    /**
     * The MR column, from merge requests alone.
     *
     * Static because the dashboard row renders the same cell and only
     * sometimes holds an Issue: a merge request whose issue is closed, or
     * outside the snapshot's queue, still has an MR column and no
     * contribution to ask. One implementation, so the two views cannot
     * describe the same merge requests differently.
     *
     * @param list<MergeRequest> $mergeRequests every MR the row covers
     * @param ?MergeRequest      $landed        the one whose work is already in
     */
    public static function renderMergeRequestCell(
        array $mergeRequests,
        ?MergeRequest $landed,
        bool $newerWorkSinceLanding,
    ): string {
        if ($mergeRequests === []) {
            return '–';
        }

        // A landing outranks everything else the cell could say. An open issue
        // whose work is already merged is the one case a maintainer cannot
        // read off the issue at all, and it is the commonest shape of a
        // Project Update Bot compatibility issue, which convention keeps open
        // so the bot can post again.
        if ($landed !== null) {
            return sprintf(
                '!%d merged %s%s',
                $landed->iid,
                substr((string) $landed->mergedAt, 0, 10),
                $newerWorkSinceLanding ? ', newer work since' : '',
            );
        }

        $substantive = self::substantive($mergeRequests);
        $representative = $substantive[0] ?? $mergeRequests[0];

        $cell = '!' . $representative->iid;
        if ($representative->carriesChanges() === false) {
            $cell .= ' empty';
        }

        $others = \count($mergeRequests) - 1;

        return $others > 0 ? $cell . ' +' . $others : $cell;
    }
}

File: go/internal/patches/contribution.go
package patches

import (
	"fmt"
	"time"

	"github.com/owenbush/upkeep/internal/drupal"
	"github.com/owenbush/upkeep/internal/gitlab"
)

// Contribution is one drupal.org issue together with every merge request that
// claims authorship of it — the unit `upkeep patches` classifies and renders.
//
// Both halves are kept because an issue can hold both kinds of contribution at
// once, which is normal in the Drupal community: a patch posted in comment 4,
// a merge request opened in comment 9, and a re-roll posted in comment 14 that
// never made it onto the branch.
type Contribution struct {
	Module string
	Issue  drupal.Issue
	// MergeRequests is every merge request whose metadata asserts authorship
	// of this issue.
	MergeRequests []gitlab.MergeRequest
}

// Pair matches each issue with the merge requests that claim authorship of it.
//
// Shared by every consumer that has to decide what a contribution *is* — the
// patch report and the dashboard — so the two can never disagree about whether
// an issue is covered.
//
// Merge requests claiming an issue outside issues are dropped as they are
// grouped. That is not observable in the result — only the given issues get a
// contribution either way — it just keeps the grouping bounded by the issues
// asked about rather than by however many merge requests the project has.
//
// forkNids maps source-project id to issue nid. It is the authoritative
// pairing where it has an answer: drupal.org made that fork *for* that issue,
// which is a fact about how the repository exists rather than a string in a
// title.
func Pair(
	module string,
	issues []drupal.Issue,
	mergeRequests []gitlab.MergeRequest,
	forkNids map[int]int,
) []Contribution {
	wanted := make(map[int]bool, len(issues))
	for _, issue := range issues {
		wanted[issue.Nid] = true
	}

	byNid := map[int][]gitlab.MergeRequest{}
	for _, mr := range mergeRequests {
		// The fork wins. It is the only thing that pairs a Project Update Bot
		// merge request at all: those are titled "Automated Project Update Bot
		// fixes" on a branch called project-update-bot-only, and say only
		// "Relates to #NNN" — which ExtractOwningIssue rejects by design, so a
		// bot cannot suppress an issue's patches by mentioning it.
		nid, found := 0, false
		if mr.SourceProjectID != 0 {
			nid, found = forkNids[mr.SourceProjectID]
		}
		if !found {
			nid, found = drupal.ExtractOwningIssue(mr.Title, mr.SourceBranch, mr.Description)
		}
		if !found || !wanted[nid] {
			continue
		}
		byNid[nid] = append(byNid[nid], mr)
	}

	contributions := make([]Contribution, 0, len(issues))
	for _, issue := range issues {
		contributions = append(contributions, Contribution{
			Module: module, Issue: issue, MergeRequests: byNid[issue.Nid],
		})
	}

	return contributions
}

// Landed is the merge request whose work has landed, if one has.
//
// The question this whole surface exists to answer for a class of issue that
// cannot answer it itself: Project Update Bot compatibility issues are kept
// open on purpose, so the bot can post again as core moves, and an open one
// may already have had its work merged months ago. Two such issues look
// identical until you ask whether anything was merged.
func (c Contribution) Landed() *gitlab.MergeRequest {
	return gitlab.LatestMerged(c.MergeRequests)
}

// HasWorkNewerThanLanding reports whether anything on the issue is newer than
// the merge.
//
// The discriminator, and the reason this never says "resolved". A bot that
// posts again after its earlier work merged has raised new work; an issue
// where nothing has happened since is one whose open status is only the
// convention. Both are true statements about evidence, and which of them
// warrants closing the issue stays the maintainer's call.
func (c Contribution) HasWorkNewerThanLanding() bool {
	landed := c.Landed()
	if landed == nil || landed.MergedAt == "" {
		return false
	}
	mergedAt, ok := parseTimestamp(landed.MergedAt)
	if !ok {
		return false
	}

	for _, file := range c.Issue.Files {
		if file.Timestamp > mergedAt.Unix() {
			return true
		}
	}
	for _, mr := range c.MergeRequests {
		if mr.State == "merged" || mr.UpdatedAt == "" {
			continue
		}
		if updated, ok := parseTimestamp(mr.UpdatedAt); ok && updated.After(mergedAt) {
			return true
		}
	}

	return false
}

// parseTimestamp reads the shapes GitLab sends. PHP's strtotime takes
// essentially anything; Go needs the accepted forms listed, so the ones the
// API actually uses come first.
func parseTimestamp(raw string) (time.Time, bool) {
	for _, layout := range []string{
		time.RFC3339Nano,
		time.RFC3339,
		"2006-01-02T15:04:05.000Z",
		"2006-01-02 15:04:05 -0700",
		"2006-01-02T15:04:05",
		"2006-01-02",
	} {
		if parsed, err := time.Parse(layout, raw); err == nil {
			return parsed, true
		}
	}

	return time.Time{}, false
}

// SubstantiveMergeRequests is the merge requests that carry changes.
//
// One whose emptiness is *unknown* counts as substantive. Unknown is the
// reading a merge-request list payload gives (it omits diff_refs) and the
// reading a snapshot cached by an older upkeep gives, and in both cases the
// safe direction is the one that preserves the pre-existing behaviour — treat
// the merge request as real work — rather than one that invents empty merge
// requests out of missing data.
func (c Contribution) SubstantiveMergeRequests() []gitlab.MergeRequest {
	return Substantive(c.MergeRequests)
}

// Substantive filters out only the merge requests *known* to be empty.
func Substantive(mergeRequests []gitlab.MergeRequest) []gitlab.MergeRequest {
	substantive := []gitlab.MergeRequest{}
	for _, mr := range mergeRequests {
		if mr.CarriesChanges() != gitlab.No {
			substantive = append(substantive, mr)
		}
	}

	return substantive
}

// Kind is how this issue's work has been delivered.
func (c Contribution) Kind() Kind {
	hasPatches := c.Issue.PatchCount() > 0
	substantive := len(c.SubstantiveMergeRequests()) > 0

	if !hasPatches {
		if substantive {
			return MergeRequestOnly
		}

		return Nothing
	}
	if substantive {
		return PatchAndMergeRequest
	}
	if len(c.MergeRequests) == 0 {
		return PatchOnly
	}

	return PatchWithEmptyMergeRequest
}

// CurrentRevision is the revision a cached check result must name to be
// current for this issue: the newest patch on it.
//
// Empty when the issue carries no patch at all, in which case there is nothing
// a result could be about.
func (c Contribution) CurrentRevision() string {
	latest, found := c.Issue.LatestPatch()
	if !found {
		return ""
	}

	return Revision(latest.URL)
}

// DashboardStatus is the STATUS cell for a patch row: what arrived, and how
// much of it.
//
// Deliberately not a gate verdict — nothing here is mergeable, and a cell that
// looked like one would invite the wrong action.
func (c Contribution) DashboardStatus() string {
	files := fmt.Sprintf("%d patches", c.Issue.PatchCount())
	if c.Issue.PatchCount() == 1 {
		files = "1 patch"
	}

	switch c.Kind() {
	case PatchOnly:
		return "PATCH " + files
	case PatchAndMergeRequest, PatchWithEmptyMergeRequest:
		return "PATCH " + files + ", " + c.MergeRequestCell()
	case MergeRequestOnly:
		return "PATCH covered by " + c.MergeRequestCell()
	default:
		return "PATCH nothing attached"
	}
}

// MergeRequestCell is the MR column: the representative merge request, flagged
// when it carries nothing, with a count of any others.
//
// A substantive merge request represents the issue in preference to an empty
// one — an issue can carry both a real branch and a bot's empty draft, and the
// real branch is the answer to "is this already in git?".
func (c Contribution) MergeRequestCell() string {
	return RenderMergeRequestCell(c.MergeRequests, c.Landed(), c.HasWorkNewerThanLanding())
}

// RenderMergeRequestCell is the MR column, from merge requests alone.
//
// A free function because the dashboard row renders the same cell and only
// sometimes holds an issue: a merge request whose issue is closed, or outside
// the snapshot's queue, still has an MR column and no contribution to ask. One
// implementation, so the two views cannot describe the same merge requests
// differently.
func RenderMergeRequestCell(
	mergeRequests []gitlab.MergeRequest,
	landed *gitlab.MergeRequest,
	newerWorkSinceLanding bool,
) string {
	if len(mergeRequests) == 0 {
		return "–"
	}

	// A landing outranks everything else the cell could say. An open issue
	// whose work is already merged is the one case a maintainer cannot read
	// off the issue at all, and it is the commonest shape of a Project Update
	// Bot compatibility issue, which convention keeps open so the bot can post
	// again.
	if landed != nil {
		since := ""
		if newerWorkSinceLanding {
			since = ", newer work since"
		}
		mergedAt := landed.MergedAt
		if len(mergedAt) > 10 {
			mergedAt = mergedAt[:10]
		}

		return fmt.Sprintf("!%d merged %s%s", landed.IID, mergedAt, since)
	}

	representative := mergeRequests[0]
	if substantive := Substantive(mergeRequests); len(substantive) > 0 {
		representative = substantive[0]
	}

	cell := fmt.Sprintf("!%d", representative.IID)
	if representative.CarriesChanges() == gitlab.No {
		cell += " empty"
	}
	if others := len(mergeRequests) - 1; others > 0 {
		cell += fmt.Sprintf(" +%d", others)
	}

	return cell
}
